Runtime errors in StarVMC/StarAgmlLib when using ROOT 6.24

After a couple of fixes, our code seems to compile fine against ROOT6.24 (See https://github.com/star-bnl/star-sw/pull/619)

However, some test jobs fail at what appears to be the geometry loading stage. The errors point to the code in `StarVMC/StarAgmlLib`. Here is an example error message:

https://github.com/star-bnl/star-sw/actions/runs/6727541469/job/18287828236?pr=619

```
In file included from /tmp/root/spack-stage/spack-stage-root-6.24.06-gbmdgvxscayjf6nqc5tatcs424pan4gs/spack-build-gbmdgvx/input_line_9:15:
In file included from /tmp/root/spack-stage/spack-stage-root-6.24.06-gbmdgvxscayjf6nqc5tatcs424pan4gs/spack-build-gbmdgvx/etc/dictpch/allHeaders.h:120:
In file included from /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/set:60:
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_tree.h:1423:13: error: no viable conversion from returned value of type 'pair<...>' to function return type 'pair<...>'
            return _Res(0, _M_rightmost());
                   ^~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_tree.h:1673:19: note: in instantiation of member function 'std::_Rb_tree<int, std::pair<const int, std::vector<TString, std::allocator<TString> > >, std::_Select1st<std::pair<const int, std::vector<TString, std::allocator<TString> > > >, std::less<int>, std::allocator<std::pair<const int, std::vector<TString, std::allocator<TString> > > > >::_M_get_insert_hint_unique_pos' requested here
            auto __res = _M_get_insert_hint_unique_pos(__pos, _S_key(__z));
                         ^
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_map.h:465:15: note: in instantiation of function template specialization 'std::_Rb_tree<int, std::pair<const int, std::vector<TString, std::allocator<TString> > >, std::_Select1st<std::pair<const int, std::vector<TString, std::allocator<TString> > > >, std::less<int>, std::allocator<std::pair<const int, std::vector<TString, std::allocator<TString> > > > >::_M_emplace_hint_unique<const std::piecewise_construct_t &, std::tuple<const int &>, std::tuple<> >' requested here
          __i = _M_t._M_emplace_hint_unique(__i, std::piecewise_construct,
                     ^
/star-sw/.sl79_gcc485/OBJ/StarVMC/StarAgmlLib/AgShape.h:70:53: note: in instantiation of member function 'std::map<int, std::vector<TString, std::allocator<TString> >, std::less<int>, std::allocator<std::pair<const int, std::vector<TString, std::allocator<TString> > > > >::operator[]' requested here
  std::vector< TString > &parList(){ return mParList[mId]; }
                                                    ^
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_pair.h:127:17: note: candidate constructor not viable: no known conversion from '_Res' (aka 'pair<std::_Rb_tree_node_base *, std::_Rb_tree_node_base *>') to 'const std::pair<std::_Rb_tree_node_base *, std::_Rb_tree_node_base *> &' for 1st argument
      constexpr pair(const pair&) = default;
                ^
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_pair.h:128:17: note: candidate constructor not viable: no known conversion from '_Res' (aka 'pair<std::_Rb_tree_node_base *, std::_Rb_tree_node_base *>') to 'std::pair<std::_Rb_tree_node_base *, std::_Rb_tree_node_base *> &&' for 1st argument
      constexpr pair(pair&&) = default;
                ^
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_pair.h:124:12: note: candidate template ignored: deduced type 'const pair<...>' of 1st parameter does not match adjusted type 'pair<...>' of argument [with _U1 = std::_Rb_tree_node_base *, _U2 = std::_Rb_tree_node_base *, $2 = void]
        constexpr pair(const pair<_U1, _U2>& __p)
                  ^
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_pair.h:150:12: note: candidate template ignored: deduced type 'pair<...>' of 1st parameter does not match adjusted type 'pair<...>' of argument [with _U1 = std::_Rb_tree_node_base *, _U2 = std::_Rb_tree_node_base *, $2 = void]
        constexpr pair(pair<_U1, _U2>&& __p)
                  ^
```

cc @klendathu2k @fgeurts

--- a/StarVMC/StarAgmlLib/AgBlock.cxx
+++ b/StarVMC/StarAgmlLib/AgBlock.cxx
@@ -98,6 +98,9 @@ void AgBlock::AddGroup( const Char_t *name )
 ;//{ mGroups.push_back( name ); }
 
 
+void AgBlock::AddCut( TString cut, double value ){ mGstpar[cut] = value; }
+
+
 // >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> WHY IS THIS COMMENTED OUT???
 void AgBlock::Inherit()
 {

--- a/StarVMC/StarAgmlLib/AgBlock.h
+++ b/StarVMC/StarAgmlLib/AgBlock.h
@@ -153,7 +153,7 @@ public:
   int numberBranches() const { return mNumberBranches; }
   void branch(){ mNumberBranches++; }
 
-  void AddCut( TString cut, double value ){ mGstpar[cut] = value; }
+  void AddCut( TString cut, double value );
   std::map<TString, double>& GetCuts() { return mGstpar; }
 
  private:

--- a/StarVMC/StarAgmlLib/AgMLExtension.cxx
+++ b/StarVMC/StarAgmlLib/AgMLExtension.cxx
@@ -13,3 +13,5 @@ AgMLExtension::AgMLExtension() : TGeoRCExtension(),
 {
   Grab();
 }
+
+void AgMLExtension::AddCut( TString cut, double value ){ mGstpar[cut] = value; }

--- a/StarVMC/StarAgmlLib/AgMLExtension.h
+++ b/StarVMC/StarAgmlLib/AgMLExtension.h
@@ -57,7 +57,7 @@ public:
 
   std::vector<AgMLScoring*> GetUserHits(){ return mHitScoring; }
 
-  void AddCut( TString cut, double value ){ mGstpar[cut] = value; }
+  void AddCut( TString cut, double value );
   std::map<TString,double>& GetCuts() { return mGstpar; }
   
  

--- a/StarVMC/StarAgmlLib/AgModule.cxx
+++ b/StarVMC/StarAgmlLib/AgModule.cxx
@@ -25,6 +25,8 @@ TDataSet *AgModule::Geom()
   else           return mGeomSet;
 };
 
+void AgModule::AddHitScoring( TString name, AgMLScoring* sc ){ mHitScoring[name] = sc; }
+
 // ---------------------------------------------------------------------------------------------
 AgBlock *AgModule::AddBlock( const Char_t *name, AgBlock *_block )
 {

--- a/StarVMC/StarAgmlLib/AgModule.h
+++ b/StarVMC/StarAgmlLib/AgModule.h
@@ -40,7 +40,7 @@ class AgModule : public AgBlock
   void SetTrackingFlag( int flag ){ mTrackingFlag = flag; }
   int  GetTrackingFlag(){ return mTrackingFlag; }
 
-  void AddHitScoring( TString name, AgMLScoring* sc ){ mHitScoring[name] = sc; }
+  void AddHitScoring( TString name, AgMLScoring* sc );
   std::map<TString,AgMLScoring*> GetHitScoring(){ return mHitScoring; }
 
  private:

--- a/StarVMC/StarAgmlLib/AgShape.cxx
+++ b/StarVMC/StarAgmlLib/AgShape.cxx
@@ -626,6 +626,8 @@ Bool_t AgShape::morphed()
   return true; // No shapes match
 }
 
+std::vector< TString > &AgShape::parList(){ return mParList[mId]; }
+
 // =============================================================================================================
 Bool_t AgShape::parameterized()
 {

--- a/StarVMC/StarAgmlLib/AgShape.h
+++ b/StarVMC/StarAgmlLib/AgShape.h
@@ -67,7 +67,7 @@ class AgShape : public TNamed, public AgParameterList<double>
   Bool_t morphed();
 
   /// Return a list of the parameters which define this shape
-  std::vector< TString > &parList(){ return mParList[mId]; }
+  std::vector< TString > &parList();
 
   /// Returns true if the shape's parameters are to be set
   /// when positioned

--- a/StarVMC/StarAgmlLib/StarAgmlStacker.cxx
+++ b/StarVMC/StarAgmlLib/StarAgmlStacker.cxx
@@ -129,6 +129,12 @@ TString StarAgmlStacker::realname( TString nname )
   return mRealnames[nname];
 }
 
+
+const std::vector<TString> &StarAgmlStacker::nicknames(const Char_t *name)
+{
+  return mNicknames[ TString(name) ];
+}
+
 // ---------------------------------------------------------------------------------------------------
 void StarAgmlStacker::Debug( const Char_t *name, const Char_t *flag )
 {

--- a/StarVMC/StarAgmlLib/StarAgmlStacker.h
+++ b/StarVMC/StarAgmlLib/StarAgmlStacker.h
@@ -35,10 +35,7 @@ class StarAgmlStacker : public TNamed
 
   virtual Bool_t SearchVolume( const AgShape &shape, const AgAttribute &attr )=0;
 			
-  const std::vector<TString> &nicknames(const Char_t *name)
-    {
-      return mNicknames[ TString(name) ];
-    }
+  const std::vector<TString> &nicknames(const Char_t *name);
 
   void Debug( const Char_t *name, const Char_t *flag="position" );
 

--- a/mgr/Dyson/Export/AgROOT.py
+++ b/mgr/Dyson/Export/AgROOT.py
@@ -687,7 +687,7 @@ class StarGeometry(Handler):
         /// Construct geometry with the specified tag, and return wrapped in a TDataSet
         static TDataSet* Construct( const char* name = "%s");
         static bool      List     ( const char* name = "%s");
-        static void      Whitelist( const char* name, int value=1 ){ whitelist[name]=value; }
+        static void      Whitelist( const char* name, int value=1 );
         static void      InitAgML ( const char* stacker="StarTGeoStacker");
         StarGeometry(){ /* nada */ };
         virtual ~StarGeometry(){ /* nada */ }
@@ -718,6 +718,8 @@ class StarGeometry(Handler):
 
         std::map<std::string,int> StarGeometry::whitelist= {{"all",1}};
         
+        void StarGeometry::Whitelist(const char* name, int value){ whitelist[name]=value; }
+
         TDataSet* StarGeometry::Construct( const char* name )
         {
         std::string tag = name;

